util/misc.py

```diff
@@ -428,6 +428,40 @@ def available_cpu_count():
 
     return multiprocessing.cpu_count()
 
+def sanitize_thread_count(threads=None, tool_max_cores_value=available_cpu_count):
+    ''' Given a user specified thread count, this function will:
+        - ensure that 1 <= threads <= available_cpu_count()
+        - interpret None values to mean max available cpus
+            unless PYTEST_XDIST_WORKER_COUNT is defined as an environment
+            variable, in which case we always return 1
+        - allow for various, tool-specific ways of specifying
+            max available cores (tool_max_value)
+        tool_max_cores_value can be one of:
```

Comment:
Wouldn't this just be a num_cores override? Aka if defined set threads = the parameter. The naming and below logic is more complicated than necessary.

Reply:
The idea is that this is an override only if the user wanted max cores, but doesn't override if someone wanted a specific value. For example usages, see tools/gap2seq (where it is set to 0) or tools/mafft (where it is set to -1). The Novoalign tool doesn't really have a --threads param yet (it probably should) but there it should be set to None.

Reply:
It's a bit weird to see `tool_max_cores_value=available_cpu_count` in the function signature since `available_cpu_count` is a function. Normally we would set the default to `None` and call the default function within the function? Especially because `tool_max_cores_value` is supposed to be an int and not a function if it's actually specified.

Reply:
Yeah I agree.. I dislike that too. None would be the obvious default, I think the only reason I didn’t use it is in the case that the user passed None in this param, it could return None (for example if we added a threads param to Novoalign, it would simply not specify anything to let the tool choose max cores by itself). I actually intended the default return value to be available_cpu_count() (ie the int return value from calling the function) but python (or lint, I forget) hates when you do that kind of thing.

I agree we should do something better here..